### PR TITLE
GEODE-7241 Refactor publish.gradle into -java and -war types

### PR DESCRIPTION
**Pull request by rhoughton-pivot**
Allows publication of Jar or War artifacts via an `apply` statement.
We then apply publish-war.gradle to our projects that build war
artifacts (and that we want to publish) and publish-java.gradle to the
jar libraries that we want to publish.

Contents of `./gradle publishToMavenLocal` with this change:

/Users/rhoughton/.m2/repository/org/apache/geode/
├── apache-geode
│   ├── 1.11.0-SNAPSHOT
│   │   ├── apache-geode-1.11.0-SNAPSHOT.pom
│   │   ├── apache-geode-1.11.0-SNAPSHOT.tgz
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-all-bom
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-all-bom-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-client-bom
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-client-bom-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-common
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-common-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-common-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-common-1.11.0-SNAPSHOT.jar
│   │   ├── geode-common-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-concurrency-test
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-concurrency-test-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-concurrency-test-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-concurrency-test-1.11.0-SNAPSHOT.jar
│   │   ├── geode-concurrency-test-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-connectors
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-connectors-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-connectors-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-connectors-1.11.0-SNAPSHOT.jar
│   │   ├── geode-connectors-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-core
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-core-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-core-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-core-1.11.0-SNAPSHOT.jar
│   │   ├── geode-core-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-cq
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-cq-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-cq-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-cq-1.11.0-SNAPSHOT.jar
│   │   ├── geode-cq-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-dunit
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-dunit-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-dunit-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-dunit-1.11.0-SNAPSHOT.jar
│   │   ├── geode-dunit-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-experimental-driver
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-experimental-driver-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-experimental-driver-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-experimental-driver-1.11.0-SNAPSHOT.jar
│   │   ├── geode-experimental-driver-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-http-service
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-http-service-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-http-service-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-http-service-1.11.0-SNAPSHOT.jar
│   │   ├── geode-http-service-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-junit
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-junit-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-junit-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-junit-1.11.0-SNAPSHOT.jar
│   │   ├── geode-junit-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-log4j
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-log4j-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-log4j-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-log4j-1.11.0-SNAPSHOT.jar
│   │   ├── geode-log4j-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-lucene
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-lucene-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-lucene-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-lucene-1.11.0-SNAPSHOT.jar
│   │   ├── geode-lucene-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-management
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-management-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-management-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-management-1.11.0-SNAPSHOT.jar
│   │   ├── geode-management-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-memcached
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-memcached-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-memcached-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-memcached-1.11.0-SNAPSHOT.jar
│   │   ├── geode-memcached-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-old-client-support
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-old-client-support-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-old-client-support-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-old-client-support-1.11.0-SNAPSHOT.jar
│   │   ├── geode-old-client-support-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-protobuf
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-protobuf-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-protobuf-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-protobuf-1.11.0-SNAPSHOT.jar
│   │   ├── geode-protobuf-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-protobuf-messages
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-protobuf-messages-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-protobuf-messages-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-protobuf-messages-1.11.0-SNAPSHOT.jar
│   │   ├── geode-protobuf-messages-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-pulse
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-pulse-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-pulse-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-pulse-1.11.0-SNAPSHOT.jar
│   │   ├── geode-pulse-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-rebalancer
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-rebalancer-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-rebalancer-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-rebalancer-1.11.0-SNAPSHOT.jar
│   │   ├── geode-rebalancer-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-redis
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-redis-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-redis-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-redis-1.11.0-SNAPSHOT.jar
│   │   ├── geode-redis-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-serialization
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-serialization-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-serialization-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-serialization-1.11.0-SNAPSHOT.jar
│   │   ├── geode-serialization-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-unsafe
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-unsafe-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-unsafe-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-unsafe-1.11.0-SNAPSHOT.jar
│   │   ├── geode-unsafe-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-wan
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-wan-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-wan-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-wan-1.11.0-SNAPSHOT.jar
│   │   ├── geode-wan-1.11.0-SNAPSHOT.pom
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-web
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-web-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-web-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-web-1.11.0-SNAPSHOT.pom
│   │   ├── geode-web-1.11.0-SNAPSHOT.war
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
├── geode-web-api
│   ├── 1.11.0-SNAPSHOT
│   │   ├── geode-web-api-1.11.0-SNAPSHOT-javadoc.jar
│   │   ├── geode-web-api-1.11.0-SNAPSHOT-sources.jar
│   │   ├── geode-web-api-1.11.0-SNAPSHOT.pom
│   │   ├── geode-web-api-1.11.0-SNAPSHOT.war
│   │   └── maven-metadata-local.xml
│   └── maven-metadata-local.xml
└── geode-web-management
    ├── 1.11.0-SNAPSHOT
    │   ├── geode-web-management-1.11.0-SNAPSHOT-javadoc.jar
    │   ├── geode-web-management-1.11.0-SNAPSHOT-sources.jar
    │   ├── geode-web-management-1.11.0-SNAPSHOT.pom
    │   ├── geode-web-management-1.11.0-SNAPSHOT.war
    │   └── maven-metadata-local.xml
    └── maven-metadata-local.xml

Authored-by: Robert Houghton <rhoughton@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [n/a] Have you written or updated unit tests to verify your changes?

- [n/a] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-pulse/build.gradle [hunk 1]
```diff
@@ -169,6 +170,8 @@ artifacts {
 }
 
 war {
+  enabled = true
+  rootSpec.exclude("**/*commons-logging-*.jar")
```
Comment (jujoramos) [inline geode-pulse/build.gradle]: Out of curiosity: why are we excluding this library from the WAR file?.

File: geode-web/build.gradle [hunk 2]
```diff
@@ -69,5 +69,10 @@ distributedTest.dependsOn(war)
 integrationTest.dependsOn(war)
 
 war {
+  enabled = true
+  rootSpec.exclude("**/*commons-logging-*.jar")
   duplicatesStrategy = DuplicatesStrategy.EXCLUDE
+  // this shouldn't be necessary but if it's not specified we're missing some of the jars
+  // from the runtime classpath
+  classpath configurations.runtimeClasspath
```
Comment (upthewaterspout) [inline geode-web/build.gradle]: Similar to Juan's question above - why did we have to why did with add the above lines to the configuration for geode-web?

**Comment by pivotal-jbarrett** [comment]
> I have a question on the discussion thread I would like addressed before reviewing this PR.

I’ve lifted my block based on the email thread outcome. I’m indifferent to this effort.

File: geode-pulse/build.gradle (hunk 1 above)
Comment (rhoughton-pivot) [reply geode-pulse/build.gradle]: I was making the `war` spec the same as our other definitions. if they *should* be different, I'm happy to drop the modification here. The problem (and reason this is a draft PR) is that nobody has definitive answers to the expectations of these artifacts :(

File: geode-web/build.gradle (hunk 2 above)
Comment (rhoughton-pivot) [reply geode-web/build.gradle]: See my comment to Juan.. I do not have answers, only potential solutions based on input here, from people with strong opinions.

File: geode-pulse/build.gradle (hunk 1 above)
Comment (jujoramos) [reply geode-pulse/build.gradle]: Fair enough. I don't have any strong opinions either, I'm just worried about breaking PULSE logging due to the removal of this particular dependency... maybe we can test that before merging the changes?.

**Pull request merged**